xonsh/tools.py

```diff
@@ -92,6 +93,85 @@ def __init__(self, returncode, command, output=None, stderr=None,
         self.completed_command = completed_command
 
 
+class EnvPath(MutableSequence):
+    """
+    A class that implements an environment path, which is a list of
+    strings. Provides a custom method that expands all paths if the
+    relevant env variable has been set.
+    """
+    @staticmethod
+    def _expandpath(path):
+        """
+        Performs environment variable / user expansion on a given path
+        if the relevant flag has been set.
+        """
+        env = getattr(builtins, '__xonsh_env__', os.environ)
+        if env.get('EXPAND_ENV_VARS', False):
+            # expand variables and use os.path.realpath to handle cases
+            # with relative paths like ../ or ./
+            path = os.path.realpath(os.path.expanduser(expandvars(path)))
+        return path
+
+    @staticmethod
+    def _decode_bytes(path):
+        """
+        Tries to decode a path in bytes using XONSH_ENCODING if available,
+        otherwise using sys.getdefaultencoding().
+        """
+        env = getattr(builtins, '__xonsh_env__', os.environ)
+        enc = env.get('XONSH_ENCODING', sys.getdefaultencoding())
+        return path.decode(encoding=enc)
+
+    def __init__(self, *args):
+        self._d = []
+        if len(args) > 0:
+            if isinstance(args[0], str):
+                self._d = args[0].split(os.pathsep)
+            elif isinstance(args[0], pathlib.Path):
+                self._d = [args[0]]
+            elif isinstance(args[0], bytes):
+                # decode bytes to a string and then split based on
+                # the default path separator
+                self._d = EnvPath._decode_bytes(args[0]).split(os.pathsep)
+            elif isinstance(args[0], list):
+                self._d = list(args[0])
+
+    def __getitem__(self, item):
+        return EnvPath._expandpath(self._d[item])
+
+    def __setitem__(self, index, item):
+        self._d.__setitem__(index, item)
+
+    def __len__(self):
+        return len(self._d)
+
+    def __delitem__(self, key):
+        self._d.__delitem__(key)
+
+    def insert(self, index, value):
+        self._d.insert(index, value)
+
+    def get_paths(self):
+        """
+        Returns the list of directories that this EnvPath contains.
+        """
+        return list(self)
+
+    def __repr__(self):
+        return repr(self._d)
+
+    def __eq__(self, other):
+        # Compare using sets to account for different
+        # order in items as well as unorderable lists
+        # (e.g. containing Path() and str())
+        if isinstance(other, list):
+            return set(self._d) == set(other)
+        elif isinstance(other, EnvPath):
+            return set(self) == set(other)
+        else:
+            return False
```

Comment:
That would result in two `EnvPath` objects that contain the same entries in different order to be considered unequal. Do we want that?
Now that I think of it, would we have to compare two `EnvPath` objects at any case? I cannot seem to think of such a scenario, so you might be right after all.